{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Downstream Applications\n",
    "\n",
    "Finally, let's apply the trained models. We will test on several cases and proceed by votes from the cross-validated models. Several applications are\n",
    "\n",
    "* Visualise binding specificity prediction on voxel as pymol session isosurface. For AUCG, we should aim at a multi-class multi-label situation, but for SXPR prediction is unlikely multilabel. These two cases are handled by levelling on the bit height (bit fraction) on logo diagrams.\n",
    "* Visualise binding site 'projected' onto protein surface. This is a popular comparing apple-to-orange scenario when users want to adapt our halo-based prediction to prediciton of site/non-site on protein sequence.  \n",
    "* Sequence Logo diagram.\n",
    "\n",
    "\n",
    "We will use 16-layer Resnet models as a baseline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "# ============== Click Please.Imports\n",
    "import sys\n",
    "import glob\n",
    "import gc\n",
    "\n",
    "\n",
    "import random\n",
    "random.seed(0)\n",
    "import pandas as pd\n",
    "\n",
    "import torch\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import glob\n",
    "import sys\n",
    "import shutil\n",
    "import tqdm\n",
    "\n",
    "\n",
    "\n",
    "# ================\n",
    "# Torch related\n",
    "# ==============\n",
    "import torch \n",
    "#from torch import nn\n",
    "\n",
    "\n",
    "\n",
    "#import pytorch_lightning as pl\n",
    "#import torchmetrics\n",
    "\n",
    "# Turn on cuda optimizer\n",
    "print(torch.backends.cudnn.is_available())\n",
    "torch.backends.cudnn.enabled = True\n",
    "torch.backends.cudnn.benchmark = True\n",
    "# disable debugs NOTE use only after debugging\n",
    "torch.autograd.set_detect_anomaly(False)\n",
    "torch.autograd.profiler.profile(False)\n",
    "torch.autograd.profiler.emit_nvtx(False)\n",
    "# Disable gradient tracking\n",
    "torch.no_grad()\n",
    "\n",
    "\n",
    "# =============\n",
    "# NN\n",
    "# =================\n",
    "sys.path.append('../')\n",
    "from NucleicNet.DatasetBuilding.util import *\n",
    "from NucleicNet.DatasetBuilding.commandReadPdbFtp import ReadBCExternalSymmetry, MakeBcClanGraph\n",
    "from NucleicNet.DatasetBuilding.commandDataFetcher import FetchIndex, FetchTask, FetchDataset\n",
    "from NucleicNet.DatasetBuilding.commandBenchmark import BenchmarkWrapper\n",
    "from NucleicNet import Burn, Fuel\n",
    "import NucleicNet.Burn.M1\n",
    "import NucleicNet.Burn.util\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "%config InlineBackend.figure_format = 'svg'\n",
    "\n",
    "sns.set_context(\"notebook\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading pdb and featurize\n",
    "\n",
    "Note that user should make sure that the pdb submitted does not contain non-protein parts and that the protein submitted is intended, even though we have script to retain only the following 20 canonical protein residues:\n",
    "`                    \"ALA\",\"CYS\",\"ASP\",\"GLU\",\"PHE\",\"GLY\", \n",
    "                    \"HIS\",\"ILE\",\"LYS\",\"LEU\",\"MET\",\"ASN\", \n",
    "                    \"PRO\",\"GLN\",\"ARG\",\"SER\",\"THR\",\"VAL\", \n",
    "                    \"TRP\",\"TYR\"` as a basic sanitization procedure, we do not replace non-canonical amino acids, which will be left as a hole if present.\n",
    "\n",
    "Also, do not submit a C-alpha only structure or a structure with no sidechain or heavily stubbed sidechain (particularly common practice in cryoEM!). For simplicity, we also do not do biological assembly and users should look out for rna binding at the interface of assembled unit proteins.\n",
    "\n",
    "* The pdb files has to be put into `../LocalServerExample/` if you are using this notebook. The output is stored in a user designated folder e.g. `../ServerExample/ServerOutput/\"`\n",
    "\n",
    "# Estimated Time of Arrival \n",
    "We do not offer CPU version anymore. Runtime with a GTX 1070 for 4f3t\n",
    "* Preparation: 10 minutes\n",
    "* Make SXPR: 2 minutes for three 16-layer resnet\n",
    "* Make AUCG: 2 minutes for three 16-layer resnet\n",
    "* Total time: 14 minutes, scale almost linearly with number of atoms.\n",
    "\n",
    "Running all the examples takes around 3-4 hours.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from NucleicNet.DatasetBuilding.commandServer import Server\n",
    "\n",
    "# NOTE This folder contains test case pdb to be sanitised.\n",
    "DIR_ServerExample = '../LocalServerExample/'\n",
    "UploadTestCases = sorted(glob.glob(\"%s/*.pdb\" %(DIR_ServerExample))) \n",
    "\n",
    "\n",
    "DIR_ServerFolder = \"../LocalServerExample/ServerOutputV1p1/\" \n",
    "MkdirList([DIR_ServerFolder])\n",
    "# NOTE Make some subfolders\n",
    "for i in range(len(UploadTestCases)):\n",
    "    try:\n",
    "        pdbid = UploadTestCases[i].split(\"/\")[-1].split(\".\")[0]\n",
    "        print(\"Working on %s\" %(pdbid))\n",
    "\n",
    "        ServerSubFolder = DIR_ServerFolder + \"%s/\"%(pdbid)\n",
    "        ServerC = Server(SaveCleansed = True, SaveDf = True,\n",
    "                        Select_HeavyAtoms = True,\n",
    "                        Select_Resname = [\n",
    "                            \"ALA\",\"CYS\",\"ASP\",\"GLU\",\"PHE\",\"GLY\", \n",
    "                            \"HIS\",\"ILE\",\"LYS\",\"LEU\",\"MET\",\"ASN\", \n",
    "                            \"PRO\",\"GLN\",\"ARG\",\"SER\",\"THR\",\"VAL\", \n",
    "                            \"TRP\",\"TYR\"\n",
    "                            ],\n",
    "                        DIR_ServerFolder = ServerSubFolder,\n",
    "                        DsspExecutable = \"../NucleicNet/util/dssp\",\n",
    "                        DIR_ClassIndexCopy = \"../Database-PDB/DerivedData/ClassIndex.pkl\")\n",
    "\n",
    "        \n",
    "        ServerC.SimpleSanitise(aux_id=0, DIR_InputPdbFile=UploadTestCases[i])\n",
    "        ServerC.MakeHalo(aux_id=0)\n",
    "        ServerC.MakeDssp(aux_id=0)\n",
    "        ServerC.MakeDummyTypi(aux_id=0)\n",
    "        ServerC.MakeFeature()\n",
    "        ServerC.MakeSXPR(aux_id=0, Checkpointlist = [   # NOTE These three models (with gelu) \n",
    "                                            \"../Models/SXPR-9CV_SXPR-9CV/132_133/checkpoints/epoch=2-step=51700-hp_metric=0.5486971139907837.ckpt\",\n",
    "                                            \"../Models/SXPR-9CV_SXPR-9CV/134_135/checkpoints/epoch=2-step=51473-hp_metric=0.5440343022346497.ckpt\",\n",
    "                                            \"../Models/SXPR-9CV_SXPR-9CV/136_137/checkpoints/epoch=2-step=52381-hp_metric=0.4679257273674011.ckpt\",\n",
    "\n",
    "                                                        \n",
    "                                                        ]\n",
    "                                                )\n",
    "        ServerC.MakeAUCG(aux_id=0,  Checkpointlist = [  # NOTE These three models (with gelu)\n",
    "                                                        \"../Models/AUCG-9CV_AUCG-9CV/294_295/checkpoints/epoch=7-step=26400-hp_metric=0.485228568315506.ckpt\",\n",
    "                                                        \"../Models/AUCG-9CV_AUCG-9CV/296_297/checkpoints/epoch=7-step=25924-hp_metric=0.48591428995132446.ckpt\",\n",
    "                                                        \"../Models/AUCG-9CV_AUCG-9CV/298_299/checkpoints/epoch=7-step=24326-hp_metric=0.5161714553833008.ckpt\",\n",
    "\n",
    "                                                        ]\n",
    "                                                )\n",
    "    except:\n",
    "        print('Cuda Runtime error in %s' %(pdbid) )\n",
    "        del ServerC\n",
    "        continue\n",
    "    # NOTE Downstream applications\n",
    "    ServerC.Downstream_VisualisePse(aux_id=0)\n",
    "\n",
    "\n",
    "    # NOTE Sequence Logo Making. We will simply take the centroid from uploaded structure.\n",
    "    try:\n",
    "        ServerC.Downstream_Logo(\n",
    "                        DIR_InputPdbFile = UploadTestCases[i],\n",
    "                        User_ProteinContactThreshold = 6.0,\n",
    "                        User_SummarisingSphereRadius = 1.5)\n",
    "    except AssertionError:\n",
    "        print(\"ABORTED. Supply template base location for %s\" %(pdbid))\n",
    "\n",
    "    del ServerC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Summary of output files\n",
    "* `uploaded.pml` is the pymol script to generate `uploaded.pse`. This pse file can be opened with pymol > 2.4 `pymol uploaded.pse` and contains visualisation of the following files. The `occupancy` column in PDB format is used to store the probability and the `b-factor` column stores the bit fraction (fractional height in sequence logos.) Visualising multi-label results on voxel is difficult as the results overlap. We provide visualisation of the multi-class multi-label result with simple heuristic filters on bit fraction (to remove ambiguous base-sites) and DBSCAN (to remove spatial outliers).\n",
    "* * `Result_0.4_Filtered.pdb` is the DBSCAN filtered result for visualisation of AUCG with bit fraction > 0.4. \n",
    "* * `Result_0.6_Filtered.pdb` is the DBSCAN filtered result for visualisation of PR with bit fraction > 0.6\n",
    "* * `Result_OrangeSitePerAtom.pdb` is the in case users are interested in 'projection' of halo-based multi-class result for a binary classification (site/non-site) done on protein sequence positions. A per-atom heuristic score is stored in the `occupancy` column; it is calculated by counting the number of voxels corresponding to site labels. In these cases of comparing apple to orange (two different interpretable task-orientated calculations), an important remark is that taking a mean of the score over a residue may not be correct as some protein residue sidechains are buried/surfaced and only interacts with backbone; similar to adenosine binding in kinases hinge. We would suggest to take a max per residue even though it may result in false positive and user should ramp for an optimal cutoff when comparing to their binary classification. A visualisation of the per-atom score is visualised as surface in the `uploaded.pse`.\n",
    "* `uploaded_*_logo_RNACColor.png` is the Sequence logo predicted by NucleicNet per base centroid location indicated by the uploaded input PDB file. `*` is the widcard for nucleotide chain ID in the file supplied by user. \n",
    "\n",
    "\n",
    "# Comments on Visualization\n",
    "* `4f3t`, `2ez6`, `3k62`. These are cases we presented in the paper with visually appealing results. Personal picks are the Argonaute 2 `4f3t` and PUF `3k62`, where all the components are precisely visualised. `2ez6` is a dsRNA complex, while the phospho-sugar backbone is nicely presented, base preference is scarce.\n",
    "\n",
    "* `7ki3` is an Argonaute 2 released Oct 2021, not in our dataset. Unlike other argonautes structure, it contains a miR-122:Site-1 three-way junction and an extra apparent binding site at G21 with loose contact with a proline (3.1 angstrom best) which the author described `despite the numerous contacts described above, SL1 and t9/G25 are both detrimental to Site-1 binding compared to a Site-1 mutant RNA with an unpaired A4 bridge between seed and supplementary paired regions. We suspect these differences may reflect tension in the Ago2:miR-122 complex, which must adopt an extended conformation to engage SL1 `. In our prediction, as expected, sites shows up satisfactorily with accurate labels, including where phosphate makes contact with the new miR-122 duplex, but in absence of specificity at G21. Note, Jordy also confirmed in email with the author Prof. Ian J. MaeRae that \"t9/G25\" is a typo in the statement above and should read \"t9/G21\".\n",
    "\n",
    "* `5z9wLess` is a large Ebola virus nucleoprotein-RNA complex. For simplicity, only 6 assembled unit of protein is retained (,though a full result is also obtainable offline in an hour using the file `5z9w.pdb`). Interestingly, without even updating the model on our server, the binding site is predicted correctly ([submission to our server this way](http://www.cbrc.kaust.edu.sa/NucleicNet/index.html) and we have not yet even update our online server since 2018!). Two notes here. \n",
    "* * (1) As the RNA sits near the interface of assembled protein units, we should anticipate reasonable structure-based prediction when a correctly assembled structure is provided. For example, [a simple download from the pdb](https://files.rcsb.org/download/5Z9W.pdb) is not necessarily correct. And, obviously, the users should only interpret results from chain `o` and `0` with caution at the \"upper\" and \"lower\" boundary where [\"symmetry mates\"](https://www.ccp4.ac.uk/MG/ccp4mg_help/symmetry.html) are still missing. \n",
    "* * (2) When adapting a multi-class prediction on 3D voxels to a binary prediction on 1D sequences, cautions have to be taken as of how to 'retrieve labels' on a sequence with a simple heuristic. While an okay heuristic as we did in 2019 is simply averaging the top label in the 30 voxels nearest to a protein residue, it does not mean the heuristic is genuine for all cases (these are quick fix with no theoretical bounds!) and certainly not for cherry picking. Another decision to be made in this apple-to-orange comparison is what distances residues stand when we consider them as interacting. \n",
    "\n",
    "* `7eni` is a CRISPR-associated endonuclease Cas9 with an alternate location at R61 near a phosphate binding site released Apr 2022. The pdb entry still lacks the PubmedID as of June 2022 and Cas9 is absent from our AUCG dataset for various reasons (e.g. Cryo-EM, specificity not fully understood, some cas system lacks specificity in majority of regions), but sites shows up satisfactorily with accurate labels including U/G preference at position 6 and the many phospho-sugar interactions at protein helix res.41-74. \n",
    "\n",
    "# Comments on Sequence Logo\n",
    "Finally, as a subsidiary application, we try to infer sequence logo diagrams based on structures, without invoking any training with sequencing data. Several PDB entries are presented including `1aud`, `1cvj` ,`2adc`,`2err`, `2g4b` , `2lec` ,`2py9`,  `4bs2`. Several further remarks.\n",
    "* For simplicity, we display all base positions indicated in the user-supplied PDB file. A caveat of this practice is that in some cases when a non-site is predicted (e.g. the 7ki3 G21 case presented above) but a non-specific interaction shows up in the supplied PDB file, the logo diagram display should be interpreted as if it is calculated with P(Base Identity | a Base is predicted), but this does not mean the position is necessarily sequence-specific. While this simplifies scenario where users really want to take a look at what base identity is predicted even if the NucleicNet mis-classified it as non-base, which seldom occurs, caution should be taken at interpretation. A cross check on this behavior can be done by inspecting the pymol pse; if there is no voxel labeled or when the voxel is labeled as `P` or `R`, then the position may not be interepreted as sequence specific.\n",
    "* Many of these structures are only available as [solution NMR-based structures](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/methods-for-determining-structure), where only a subset of atom coordinates are inferred from distance proxies e.g. MDS and the rest are \"guesses\" from forcefield minimization; this results in multiple satisfiable conformations, often more than 20. Not all of these conformers in the PDB files are necessarily precise. Examples of NMR-based PDB structures include `1aud`, `2adc`/`2ad9`, `2leb`/`2lec`, `4bs2`.  There are also structures with multiple chains. Several 'good-standing' states(0-indexing)/chains are listed below. \n",
    "\n",
    "* * U2AF2: `2g4b` chain B; crystal structure.\n",
    "* * PABPC: `1cvj` any chain; crystal structure.\n",
    "* * PCBP2: `2py9` chain F; crystal structure.\n",
    "* * RBFOX1: `2err` state 0\n",
    "* * SNRPA: `1aud` state 0\n",
    "* * PTBP1: `2ad9` state 1 or  `2adc` state 8 chain B; state 12/13 chain C; \n",
    "* * SRSF2: `2leb` state 7; `2lec` state 12. \n",
    "* * TARDBP: `4bs2` state 9 \n",
    "\n",
    "# Epilogue \n",
    "We have presented several downstream applications. Some further remarks.\n",
    "* Ensembles. To avoid re-doing the training with even larger memory burden and hours, we simply take the three cross-validated models and let them vote for an answer. But, of course, basic [boosting/bagging type of fine-tuned training](https://en.wikipedia.org/wiki/Boosting_(machine_learning)) can be done if higher accuracy is desired. \n",
    "* Sanitisation of input. For simplicity, in the downstream applications, we only take coordinates with the alternative locaton `A` or an empty alternative location indicator. If user want an alternative location `B`, they should manually edit their PDB file or [try to do it with pymol](https://pymolwiki.org/index.php/Property_Selectors) if possible (and think thrice! Some alternative locations can have much poorer B factor/ are simply sidechain outliers such that authors put it in the dustbin.).\n",
    "* Voting at spatial neighborhood. We found this technique (smoothing labels at spatially adjacent voxel) very useful in both basic training and post-processing. In the future, a message passing type of training on features can be done.\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "483da45aec43309c24cf49001fd3be9352115ff3f8b01e51a9517882487d0ea9"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
